### ISSUE TITLE
kl_weight in scVI for more than 20,000 cells as input

### ISSUE THREAD

**Issue by FlorianBarkmann** [scverse/scvi-tools#1569]
Hi all,

thanks for providing such a great tool! I think there is a problem in scVI with handling n_epochs_kl_warmup. If the number of cells in the AnnData is larger than 20,000,the maximum number of epochs defaults to a value smaller than n_epochs_kl_warmup. This basically turns scVI into a betaVAE with beta < 1. The problem is that in UnsupervisedTrainingMixin it is never validated that the n_epochs_kl_warmup is smaller than max_epochs.

Code to reproduce the issue (runs ~10 minutes with a GPU):

```import numpy as np
import scvi
import anndata
import pytest
import scvi.train._trainingplans

_compute_kl_weight = scvi.train._trainingplans._compute_kl_weight


def kl_weight(*args, **kwargs):
    global weight
    kl_weight = _compute_kl_weight(*args, **kwargs)
    weight = kl_weight
    return kl_weight


def test_linear_annealing(monkeypatch):
    monkeypatch.setattr("scvi.train._trainingplans._compute_kl_weight", kl_weight)
    adata = anndata.AnnData(X=np.ones((25_000, 2)))
    scvi.model.SCVI.setup_anndata(adata)
    model = scvi.model.SCVI(adata)
    model.train()
    assert np.isclose(weight, 1.0)
```

This problem is easily fixed by setting the number of n_epochs_kl_warmup to min(n_epochs_kl_warmup, max_epochs). This should be done in the train method of UnsupervisedTrainingMixin because we need access to the batch size and the number of cells to calculate max steps to also ensure that n_steps_kl_warmup is smaller than max steps.

I am happy to provide a PR that fixes this.

Best,
Florian

**Comment by adamgayoso** [comment]
Hi Florian,

This is something we are aware of, and we are actively assessing ways to remove this sort of schedule. Unfortunately as simple of a change this is, it would not be backwards compatible, so it's hard for us to add easily.

@PierreBoyeau any thoughts?

**Comment by PierreBoyeau** [comment]
Isotropic normal priors are actually not that non-informative, and may overregularize latent representations. This raises two discussions that I think we need to consider in parallel: 
1. If you use N(0,1) as your prior, what default training procedure should we use, that makes sense for most users of scvi? 
2. Should we allow users to consider  alternative priors?

1. When keeping such priors, discussions in our lab seem to indicate that a beta vae regime produces more useful representations for clustering and visualization. At the same time, they may be detrimental to DE analyses.
For this reason, I think a reasonable solution would be to warn users that scvi will not reach beta=1 when you start training and precise that you can use n_kl_epochs_warmup to override this behavior.

2. When it comes to alternative prior choices, we are working on several ideas in parallel internally. For instance, I have a hyperspherical prior parameterization rusting somewhere. That being said, I don't know how important allowing several prior choices to the codebase would be.

**Comment by FlorianBarkmann** [comment]
Hi Pierre, Hi Adam,

thanks for your responses! We have actually observed a similar effect of overregularization while working with just malignant cells. Therefore, I think having beta explicitly customizable would be a great feature.  

Regarding backward compatibility, how about we introduce beta as a parameter in training_plan it defaults to none and then you get the current behavior. If beta is set the maximum kl_weight will be beta and is reached after min(max_epochs, n_kl_epochs_warmup), with a warning, if n_kl_epochs_warmup is larger than max_epochs.

For users where the case occurs that n_epochs_kl_warmup is greater than max_epochs, we raise a deprecation warning that they should set beta explicitly. Then in future versions, beta can default to 1 without breaking people's code.

PS: regarding hyperspherical prior parameterization: https://www.nature.com/articles/s41467-021-22851-4

**Comment by adamgayoso** [comment]
@FlorianBarkmann This sounds good to me, although I would call it `max_kl_weight`. Would you be able to make a PR soon? We can get this in the next 0.17 release next week.

**Comment by adamgayoso** [comment]
Likewise, if you're willing, you could also expose `min_kl_weight` which is already implemented behind the scenes. I would also default `max_kl_weight=1` and `min_kl_weight=0`

**Comment by adamgayoso** [comment]
Sorry, just to summarize my points:

1. Add `max_kl_weight` and `min_kl_weight` to training plans
2. Use the same schedule as we currently have, but max out at `max_kl_weight` (and start at `min_kl_weight`)
3. Raise a warning that the schedule max step/epoch is less than the max number of epochs, allowing users to set it themselves.
4. If `n_steps_kl_warmup` and `n_epochs_kl_warmup` are None, default to `max_kl_weight`, which is 1 by default

**Comment by wmacnair** [comment]
Hi, I had a question related to @PierreBoyeau's point above on choice of priors.

I've been working with integrating cells from a single cell type, so that clusters represent cell _states_ rather than cell types, and the differences between them are therefore more subtle.

Running `scVI`, my experience has been that the isotropic normal prior seems to dominate the embedding. In this plot, the layout is consistent with a Gaussian blob:
![image](https://github.com/scverse/scvi-tools/assets/32297936/4a7ff8ef-ad4f-4c67-ad3a-66e110f86379)
This is exactly the same data plotted with `Harmony`:
![image](https://github.com/scverse/scvi-tools/assets/32297936/92c7ff0f-6034-4142-a46e-f6158970f3dc)
I have looked through the `Harmony` clusters and I'm pretty sure these are biologically meaningful and not spurious.

My first thought was whether using a weaker prior, e.g. N(0, 10), could help. It feels like this would at least be simple to implement. I'd be curious to hear about other options for alternative priors, and whether this was developed further.

(And please let me know if there are better places for this discussion.)

Thanks!
Will

**Comment by justjhong** [comment]
Hi @wmacnair,

Thanks for the followup. The Harmony plot you showed certainly seems to be capturing some different signal, but have you tried confirming that the scVI latent is collapsing to the prior? E.g. checking if the harmony cluster labels are randomly distributed in the scVI latent, or seeing what the distributions of the individual latent variables look like. UMAP isn't the best way to diagnose posterior collapse since it can significantly distort the distances between points.

Also, what how many epochs of kl warmup did you use in this case? The warmup can make a big difference in the model fit.

**Comment by wmacnair** [comment]
Hi @justjhong 

Thanks for getting back so quickly. 

Here is a confusion matrix of the two clusterings, showing how the `scVI` clusters are allocated to the `harmony` clusters:
![hmny_vs_scvi_2023-03-08_log_p_cl2](https://github.com/scverse/scvi-tools/assets/32297936/522fd44c-0696-4f53-8360-c65305ee030d)
So there clearly is some common structure, but also a lot of disagreement.

Could this be partial posterior collapse?

Regarding kl warmups, I just used defaults. Due to the number of cells, it seems like the full number of kl_warmups wasn't achieved - the number of epochs actually trained was 150, so I got warnings like "max_epochs=149 is less than n_epochs_kl_warmup=400". Would you expect this to make a big difference?

Will

**Comment by cane11** [comment]
I think this discussion is not in the same line as the previous discussion. It's more appropriate in our discourse forum, where we collect user questions and suggestions.
Harmony tends to produce those concise clusters. Those can be an artifact of the method or can be helpful in data analysis. Nevertheless, a meaningful metric for the scVI latent space here is to check whether a kNN can predict the harmony clusters in the scVI latent space or to run scANVI with the harmony clusters as input clusters.

**Comment by wmacnair** [comment]
Hi @cane11, ok I will post something in the discourse forum :)

For anyone following this thread, this is a plot of the same data with 500 epochs, including 400 kl_warmup epochs, which doesn't look very different:
![scvi_umap_2023-04-12_std](https://github.com/scverse/scvi-tools/assets/32297936/3110f61c-c95d-4998-a1b2-f76b6264c92c)

Will

**Comment by wmacnair** [comment]
I have now run `scANVI` using the `Harmony` clusters as labels, and there is a tiny bit more structure than before, but not much. How should I interpret this? What is the "meaningful metric" here?

More generally, it's theoretically possible for two different embedding methods to find latent spaces that are both internal consistent, but are pretty much independent of each other. (Although this would require them not to be constrained by trying to maximise the variance explained, otherwise you wouldn't get independence.) What does information being shared / not shared across latent spaces tell us?

(As an aside, you wrote "Harmony tends to produce those concise clusters. Those can be an artefact of the method or can be helpful in data analysis." Can't you say exactly the same about `scVI`? "scVI tends to produce these Gaussian blobs. Those can be an artefact of the method, or can indicate that there is truly little biological signal.")

Thanks
Will
